### PR TITLE
Replace recursive status cache with a single global fast status cache

### LINKED ISSUE #2497: Signature duplicate check with checkpointing is slow
#### Problem

Even with bloom filters, #2289, the signature filtering needs to traverse a bunch of checkpoints per signature to check if its present or not.   It looks like a single bloom filter check is going to be on the order of 1us, so 32 checks is going to be to slow for our target tps (although they can be run in parallel).

#### Proposed Solution

```
type ForkId = u64;
//probably a small amount of entries in this map, so a vec might be better
type ForkStatusMap = HashMap<ForkId, Result<()>>;
type SignatureMap = HashMap<Signature, ForkStatusMap>;
//the LastId is also from a specific fork.  But any descendants could have signatures for it.
type StatusCache = HashMap<LastId, (ForkId, SignatureMap)>;
```

* Use a single LastId delimited map that keeps track of which fork the signature had been executed in.  This data structure doesn't checkpoint, or merge, and is a single data structure for all the forks in the system.
* When checking for dup signatures in the current fork, find the results in the above data structure for all the forks a signature has been processed.  Check if the forks are an ancestor of the current fork.
* Garbage collect as LastIds expire, since they expire globally for all the forks.
*  At a specific tick count, multiple LastIds would expire, because the values are different in each fork.

* Some interesting optimizations,
```
//bitfield if the signature is present
//offset in the field is number of forks above the LastIds ForkId
type ForkStatusMap = u32;
```
and using 80 bits of the Signature instead of the full 256.
tag: @garious @rob-solana @sakridge

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Recursive status cache is to slow (@sakridge).

#### Summary of Changes

Use a single data structure to minimize the number of lookups.

Fixes #2497 

tag: @rob-solana @sakridge

### CODE REVIEW

File: runtime/src/status_cache.rs [hunk 1]
```diff
@@ -9,403 +9,229 @@ use std::ops::Deref;
...
 
-/// Each cache entry is designed to span ~1 second of signatures
-const MAX_CACHE_ENTRIES: usize = solana_sdk::timing::MAX_HASH_AGE_IN_SECONDS;
-
-type FailureMap<T> = HashMap<Signature, T>;
-
-#[derive(Clone)]
-struct Status<T> {
+// Blockhash found in the transaction
+type TransactionBlockHash = Hash
```
Comment (garious) [inline runtime/src/status_cache.rs]: `Blockhash` is one word. And no need to qualify a type with the name of one container it might be found in.

Comment (aeyakovenko) [reply runtime/src/status_cache.rs]: type alias seems better than a comment

File: runtime/src/status_cache.rs [hunk 2]
```diff
@@ -1,411 +1,261 @@
...
+impl<T: Clone> StatusCache<T> {
+    /// Check if the signature from a transaction is in any of the forks in the ancestors set.
+    pub fn get_signature_status(
+        &self,
+        sig: &Signature,
+        transaction_blockhash: &Hash,
+        ancestors: &HashMap<ForkId, usize>,
+    ) -> Option<(ForkId, T)> {
+        self.cache
+            .get(transaction_blockhash)
```
Comment (garious) [inline runtime/src/status_cache.rs]: Did you know you can use `?` on Option return types?

File: runtime/src/status_cache.rs [hunk 3]
```diff
@@ -1,411 +1,261 @@
...
-        self.merges.push_front(Arc::new(merge.unwrap()));
-        if self.merges.len() > MAX_CACHE_ENTRIES {
-            self.merges.pop_back();
-        }
+    /// Insert a new signature for a specific fork.
+    pub fn insert(&mut self, transaction_blockhash: &Hash, sig: &Signature, fork: ForkId, res: T) {
+        let sig_map = self
+            .cache
+            .get_mut(transaction_blockhash)
+            .expect("new map should be created with register_hash");
```
Comment (carllin) [inline runtime/src/status_cache.rs]: Could other validators potentially send/forward you a blockhash for a slot on a different fork that you haven't observed?

Comment (carllin) [reply runtime/src/status_cache.rs]: nvm, I see that should be filtered out by the check_age() check earlier in the pipeline

Comment (carllin) [reply runtime/src/status_cache.rs]: I'm seeing this assert on testnet. Is it possible that this insert check happens between these two lines: https://github.com/solana-labs/solana/blob/master/runtime/src/bank.rs#L433-L434, such that the check_age filter earlier in the pipeline passes, but then this assert is still triggered?

Comment (aeyakovenko) [reply runtime/src/status_cache.rs]: That’s unexpected.  I can get rid of the assert.   The latest fork referencing the block hash can be the fork id for the map.

Comment (garious) [reply runtime/src/status_cache.rs]: Did you guys know that a panic while holding a read or or write guard will cause cascading panics to anyone else waiting on that lock?

Comment (aeyakovenko) [reply runtime/src/status_cache.rs]: @garious yea, i know.  which is pretty cool way to track who is holding locks.

File: runtime/src/bank.rs [hunk 4]
```diff
@@ -793,10 +798,13 @@ impl Bank {
         &self,
         signature: &Signature,
     ) -> Option<(usize, Result<()>)> {
-        let parents = self.parents();
-        let mut caches = vec![self.status_cache.read().unwrap()];
-        caches.extend(parents.iter().map(|b| b.status_cache.read().unwrap()));
-        StatusCache::get_signature_status_all(&caches, signature)
+        let mut ancestors = HashMap::new();
```
Comment (rob-solana) [inline runtime/src/bank.rs]: copy pasta

**Pull request merged**